### PR TITLE
Added Rabin Karp String Matching in c++

### LINKED ISSUE #1130: Rabin Karp String Matching algorithm in C++

### PR DESCRIPTION
**Pull request by Anshuvats**
Fixes #1130 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Sample Input / Output.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:

- Added Implementation for Rabin Karp String Matching Algorithm with proper comments and 
 indentation in C++ programming languages

#### Languages Used:
- C++

#### Files Added:
- Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp

### CODE REVIEW

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+/*Time-Complexity:-
+The average case running time of the Rabin-Karp algorithm = O(n+m).
+The best case running time of the Rabin-Karp algorithm = O(n+m).
+The worst-case running time of the Rabin-Karp algorithm = O(nm).
+*/
+#include<iostream>
+#include<cstring>
+#define AN 256    //no. of character in alphabet
+using namespace std;
+void search(char *pattern, char *text, int PN)
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Please give a blank line before function statement.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+h : d ^ m-1
+m : pattern length
+q: any prime no given input from main
+*/
+/*Time-Complexity:-
+The average case running time of the Rabin-Karp algorithm = O(n+m).
+The best case running time of the Rabin-Karp algorithm = O(n+m).
+The worst-case running time of the Rabin-Karp algorithm = O(nm).
+*/
+#include<iostream>
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Please give a blank line before include statements.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+using namespace std;
+void search(char *pattern, char *text, int PN)
+{
+    int patlen = strlen(pattern);   //pattern length
+    int txtlen = strlen(text);      //text length
+    int val = 1;
+    int hpat = 0;      // pattern hash value
+    int htxt = 0;      //text hash value
+    int i, j;
+
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: No need of a blank line

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    for(i = 0; i < patlen - 1; i++)
+        val = (val * AN) % PN;       //PN is a prime number
+
+
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Please remove unnecessary blank lines.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    for(i = 0; i < patlen; i++)
+    {
+        hpat = (AN * hpat + pattern[i]) % PN;
+        htxt = (AN * htxt + text[i]) % PN;
+    }
+
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Please remove the blank line

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            for(j = 0; j < patlen; j++)
+            {
+                if(text[i + j] != pattern[j])
+                {
+                   break;
+                }
+            }
+            if(j == patlen)
+            {
+            cout << "Found at index starting from " << i << endl;
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Please fix indentation.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            }
+
+        }
+       //Calculate the new pattern including i+m
+       htxt = (AN * (htxt - val * text[i]) + text[i + patlen]) % PN;
+
+
+
+       if(htxt < 0){
+         htxt = htxt + PN;
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Please fix indentation

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            cout << "Found at index starting from " << i << endl;
+            }
+
+        }
+       //Calculate the new pattern including i+m
+       htxt = (AN * (htxt - val * text[i]) + text[i + patlen]) % PN;
+
+
+
+       if(htxt < 0){
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Space before {

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+         htxt = htxt + PN;
+       }
+    //cout<<p<<" , "<<t<<endl;
+
+    }
+}
+int main(){
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Blank line before main block

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    char *text = "XWINGO XOR WINGO";
+    char *pattern = "WINGO";
+    int PN = 101;  // A prime number
+    search(pattern, text, PN);
+
+//    getchar();
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Remove line 71 and 72

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    return 0;
+}
+/*Input:-XWINGO XOR WINGO
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: A blank line before Sample Comments.

**Comment by Anshuvats** [comment]
Hey @MastersAbh , please review this.

**Comment by Anshuvats** [comment]
@jainaman224  kindly look into this, please.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+void search(char *pattern, char *text, int PN)
+{
+    int patlen = strlen(pattern);   //pattern length
+    int txtlen = strlen(text);      //text length
+    int val = 1;
+    int hpat = 0;      // pattern hash value
+    int htxt = 0;      //text hash value
+    int i, j;
+    for(i = 0; i < patlen - 1; i++)
```
Comment (jainaman224) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Add space after `for`, `if`, `while` etc.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+                {
+                   break;
+                }
+            }
+            if(j == patlen)
+            {
+               cout << "Found at index starting from " << i << endl;
+            }
+        }
+       //Calculate the new pattern including i+m
```
Comment (jainaman224) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Indentation error.

**Comment by Anshuvats** [comment]
Hey, @jainaman224  I have updated this. kindly look into this.

**Comment by Anshuvats** [comment]
@jainaman224! please review into this.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+                cout << "Found at index starting from " << i << endl;
+            }
+        }
+        //Calculate the new pattern including i+m
+        htxt = (AN * (htxt - val * text[i]) + text[i + patlen]) % PN;
+        if (htxt < 0)
+        {
+            htxt = htxt + PN;
+        }
+        //cout<<p<<" , "<<t<<endl;
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: This line can be removed.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 15]
```diff
@@ -0,0 +1,76 @@
+//Rabin Karp Algorithm
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Add a space after //

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 16]
```diff
@@ -0,0 +1,76 @@
+//Rabin Karp Algorithm
+/*
+AN : range or No of alphabets
+val : AN ^ patlen-1
+patlen : pattern length
+PN: any prime no given input from main
+*/
+/*Time-Complexity:-
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: A space or endline after /*

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+*/
+/*Time-Complexity:-
+The average case running time of the Rabin-Karp algorithm = O(n+m).
+The best case running time of the Rabin-Karp algorithm = O(n+m).
+The worst-case running time of the Rabin-Karp algorithm = O(nm).
+*/
+
+#include<iostream>
+#include<cstring>
+#define AN 256    //no. of character in alphabet
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: space after //

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+        htxt = (AN * (htxt - val * text[i]) + text[i + patlen]) % PN;
+        if (htxt < 0)
+        {
+            htxt = htxt + PN;
+        }
+        //cout<<p<<" , "<<t<<endl;
+    }
+}
+
+int main(){
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Add a space before {

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp (hunk 16 above)
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: A space or endline after /*

**Comment by Anshuvats** [comment]
Hey @MastersAbh, Kindly look into this.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+/*Input:-XWINGO XOR WINGO
+         WINGO
+         101
+ OUTPUT:-Found at index starting from 1
+         Found at index starting from 11
+*/
+
+
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: Please remove one extra blank line from end of the file. One blank line is sufficient.

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+int main() {
+    char *text = "XWINGO XOR WINGO";
+    char *pattern = "WINGO";
+    int PN = 101;  // A prime number
+    search(pattern, text, PN);
+    return 0;
+}
+
+/*Input:-XWINGO XOR WINGO
```
Comment (MastersAbh) [inline Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: @Anshuvats Add a space after /*

File: Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp (hunk 16 above)
Comment (MastersAbh) [reply Rabin_Karp_String_Matching_Algo/Rabin_Karp_String_Matching.cpp]: @Anshuvats Add a space after /*

**Comment by Anshuvats** [comment]
Hey@MastersAbh, I have updated this. Kindly look into this.

**Comment by jainaman224** [comment]
Change file and folder name to Rabin_Karp.

**Comment by Anshuvats** [comment]
@jainaman224, I have changed the file name and folder name. Kindly look into this.

**Pull request merged**